examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,624 @@
+est = RouterConsumerClassifier(
+    estimator=ExampleClassifier().fit_requests(foo="aliased_foo")
+).fit_requests(foo=True)
+est.get_metadata_request()["fit"]
+
+# %%
+# Alias only on the meta-estimator. This example raises an error since there
+# will be two conflicting values for routing ``foo``.
+est = RouterConsumerClassifier(
+    estimator=ExampleClassifier().fit_requests(foo=True)
+).fit_requests(foo="aliased_foo")
+try:
+    est.get_metadata_request()["fit"]
+except ValueError as e:
+    print(e)
+
+
+# %%
+# Simple Pipeline
+# ---------------
+# A slightly more complicated use-case is a meta-estimator which does something
+# similar to the ``Pipeline``. Here is a meta-estimator, which accepts a
+# transformer and a classifier, and applies the transformer before running the
+# classifier.
+
+
+class SimplePipeline(ClassifierMixin, BaseEstimator):
+    _required_parameters = ["estimator"]
+
+    def __init__(self, transformer, classifier):
+        self.transformer = transformer
+        self.classifier = classifier
+
+    def fit(self, X, y, **fit_params):
+        metadata_request_factory(self).fit.validate_metadata(kwargs=fit_params)
+
+        transformer_fit_params = metadata_request_factory(
+            self.transformer
+        ).fit.get_method_input(ignore_extras=True, kwargs=fit_params)
+        transformer_transform_params = metadata_request_factory(
+            self.transformer
+        ).transform.get_method_input(ignore_extras=True, kwargs=fit_params)
+        self.transformer_ = clone(self.transformer).fit(X, y, **transformer_fit_params)
+        X_transformed = self.transformer_.transform(X, **transformer_transform_params)
+
+        classifier_fit_params = metadata_request_factory(
+            self.classifier
+        ).fit.get_method_input(ignore_extras=True, kwargs=fit_params)
+        self.classifier_ = clone(self.classifier).fit(
+            X_transformed, y, **classifier_fit_params
+        )
+        return self
+
+    def predict(self, X, **predict_params):
+        metadata_request_factory(self).predict.validate_metadata(kwargs=predict_params)
+
+        transformer_transform_params = metadata_request_factory(
+            self.transformer
+        ).transform.get_method_input(ignore_extras=True, kwargs=predict_params)
+        X_transformed = self.transformer_.transform(X, **transformer_transform_params)
+
+        classifier_predict_params = metadata_request_factory(
+            self.classifier
+        ).predict.get_method_input(ignore_extras=True, kwargs=predict_params)
+        return self.classifier_.predict(X_transformed, **classifier_predict_params)
+
+    def get_metadata_request(self):
+        router = (
+            MetadataRouter()
+            .add(
+                self.transformer,
+                mapping={
+                    "fit": ["fit", "transform"],
+                    "predict": "transform",
+                },
+                overwrite="smart",
+                mask=True,
```

Comment:
OK so I need to think a bit more about this masking because I did not really get it.